### PR TITLE
Added LayerNorm implementation.

### PR DESCRIPTION
**Pull request by malicd**
Added `LayerNorm`, which follows [PyTorch LayerNorm](https://pytorch.org/docs/stable/generated/torch.nn.LayerNorm.html#torch.nn.LayerNorm) conventions. The layer can be used in the same manner as `BatchNorm` layer, e.g.:
```
nn.Sequential(...
    spnn.Conv3d(...),
    spnn.LayerNorm(...),
    spnn.ReLU(True),
    ...)
```

### CODE REVIEW

**Review by zhijian-liu — APPROVED** [review]

**Comment by zhijian-liu** [comment]
Thanks for the PR!

**Comment by victoryc** [comment]
I am not sure whether this implementation of LayerNorm is really correct. PyTorch's LayerNorm, which this calls, expects the input to be of the form (N, *) where the first dimension corresponds to the batch. But, in the case of SparseTensor, the first dimension corresponds to the first dimension of "feats" corresponds to the concatenated list of non-empty voxels from all the samples/images in the batch, where the number of non-empty voxels in different images can be different.

My own realization of this issue came after I coded up GroupNorm following the same pattern as this implementation of LayerNorm. But, with that implementation, the loss was not going down as training progressed, as typically expected. Then, I noticed that PyTorch's GroupNorm function expects the input to be in the format (N, C, *) where N is the number of samples/images in the batch, and C is the number of channels. But, "feats" is not in that format.

**Comment by zhijian-liu** [comment]
@victoryc, thanks for bringing this up. I think in this case, we might not be able to naively call the native PyTorch implementation of these normalization layers. For GroupNorm, is it possible to treat the N-dimension as the spatial dimension (assuming `feats` is in the shape of NxC) and set the batch size to 1?

**Comment by victoryc** [comment]
Not sure I understand your proposal adequately. Can you please elaborate a bit?


The way I saw it, one way to correctly implement GroupNorm (GN) for sparse tensors would be as follows. Suppose we maintain the indices to the location in "feats" where each sample in the batch begins. So, if there are are k images/samples in a batch, we will maintain k integers (or, for convenience, k+1 integers, where the last value points to the end). With that, whenever we need to per sample based normalization, such as GN or LN, we can easily pick the feats corresponding to that sample and call PyTorch's GN/LN/... for those. The negative about this approach is that we would lose some efficiency. Though we would still be using the batch for other operations such as convolution, we would be doing per image operations while doing normalization.

Another option that comes to my mind is to have a custom implementation of GN/LN/... specific to sparse tensors, without relying on PyTorch's implementations of those. The con about this is that it would be more work.

**Comment by malicd** [comment]
I am not entirely sure I understand the problem here. Torch LN implementation expects input of size `(N, *)`; however `N` does not have to be the batch size but [`normalized_shape`](https://pytorch.org/docs/stable/generated/torch.nn.LayerNorm.html#torch.nn.LayerNorm). Normalization is than performed on per-sample basis. As long as the `*.F` returns *only* the features of the respective voxels this should work fine.

As far as I can tell, same thing can be applied to the GN.

Please correct me if I am wrong.

PS.

>  But, in the case of SparseTensor, the first dimension of "feats" corresponds to the concatenated list of non-empty voxel coordinates from all the samples/images in the batch, where the number of non-empty voxels in different images can be different.

Do you talk about `*.F` (features) or `*.C` (coordinates) here?

**Comment by victoryc** [comment]
If you look at the definition of LayerNorm or GroupNorm (e.g., in [this](https://arxiv.org/pdf/1803.08494.pdf) paper), you will see that they are defined for each sample in a batch. So, unless we somehow separate the features corresponding to each sample, there is no way to compute them. As it stands, "feats" in SparseTensor is concatenated result of features from different samples.

**Comment by zhijian-liu** [comment]
@victoryc and @malicd, thank you both for the discussion. @malicd, could you please modify the LayerNorm implementation according to @victoryc's GroupNorm implementation? Thanks!

**Comment by malicd** [comment]
I think @victoryc has a good point and this implementation is not correct. I also agree with his implementation idea. @zhijian-liu maybe just remove it for now? Implementation overhead is much larger than anticipated and I am not sure if I have time/need this.
